src/psyclone/psyir/nodes/array_reference.py

```diff
@@ -0,0 +1,276 @@
+from psyclone.psyir.nodes.ranges import Range
+from psyclone.psyir.symbols import DataSymbol, ScalarType
+from psyclone.errors import GenerationError
+from psyclone.psyir.nodes import BinaryOperation, Literal
+
+
+class ArrayReference(Reference):
+    '''
+    Node representing an Array reference. As such it has a reference and a
+    subscript list as children 0 and 1, respectively.
+
+    '''
+    # Textual description of the node.
+    _children_valid_format = "[DataNode | Range]*"
+    _text_name = "ArrayReference"
+
+    @staticmethod
+    def _validate_child(position, child):
+        '''
+        :param int position: the position to be validated.
+        :param child: a child to be validated.
+        :type child: :py:class:`psyclone.psyir.nodes.Node`
+
+        :return: whether the given child and position are valid for this node.
+        :rtype: bool
+
+        '''
+        # pylint: disable=unused-argument
+        return isinstance(child, (DataNode, Range))
+
+    @staticmethod
+    def create(symbol, children):
+        '''Create an ArrayReference instance given a symbol and a list of Node
+        array indices.
+
+        :param symbol: the symbol that this array is associated with.
+        :type symbol: :py:class:`psyclone.psyir.symbols.DataSymbol`
+        :param children: a list of Nodes describing the array indices.
+        :type children: list of :py:class:`psyclone.psyir.nodes.Node`
+
+        :returns: an ArrayReference instance.
+        :rtype: :py:class:`psyclone.psyir.nodes.ArrayReference`
+
+        :raises GenerationError: if the arguments to the create method \
+            are not of the expected type.
+
+        '''
+        if not isinstance(symbol, DataSymbol):
+            raise GenerationError(
+                "symbol argument in create method of ArrayReference class "
+                "should be a DataSymbol but found '{0}'.".format(
+                    type(symbol).__name__))
+        if not isinstance(children, list):
+            raise GenerationError(
+                "children argument in create method of ArrayReference class "
+                "should be a list but found '{0}'."
+                "".format(type(children).__name__))
+        if not symbol.is_array:
+            raise GenerationError(
+                "expecting the symbol to be an array, not a scalar.")
+        if len(symbol.shape) != len(children):
+            raise GenerationError(
+                "the symbol should have the same number of dimensions as "
+                "indices (provided in the 'children' argument). "
+                "Expecting '{0}' but found '{1}'.".format(
+                    len(children), len(symbol.shape)))
+
+        array = ArrayReference(symbol)
+        array.children = children
+        for child in children:
+            child.parent = array
+        return array
+
+    def __str__(self):
+        result = ("ArrayReference" + super(ArrayReference, self).__str__() +
```

Comment:
This currently prints the node name as "ArrayReferenceArrayReference", please remove the first hardcoded string. There are a few tests that needs fixing after this.

Reply:
Oh yes. Embarrassingly, I remember thinking that looked weird when I did the search-n-replace but I didn't do anything about it!